File: backend/src/mindloom/services/runs.py
import uuid
from datetime import datetime
from typing import Optional, Dict, Any

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from mindloom.app.models.run import RunORM, RunStatus
# TODO: Adjust if your Pydantic create schema is named differently or located elsewhere
# from mindloom.app.schemas.run import RunCreate 

class RunService:
    """Service layer for Run related operations."""

    async def create_run(
        self,
        db: AsyncSession,
        *,
        runnable_id: uuid.UUID,
        runnable_type: str,
        input_variables: Optional[Dict[str, Any]] = None,
        user_id: Optional[uuid.UUID] = None # Optional: if you track which user initiated the run
    ) -> RunORM:
        """
        Creates a new run record in the database with PENDING status.

        Args:
            db: The AsyncSession for database interaction.
            runnable_id: The UUID of the agent or team being run.
            runnable_type: The type of runnable ('agent' or 'team').
            input_variables: The input data for the run.
            user_id: The ID of the user initiating the run (optional).

        Returns:
            The created RunORM object.
        """
        db_run = RunORM(
            runnable_id=runnable_id,
            runnable_type=runnable_type,
            input_variables=input_variables,
            user_id=user_id,
            status=RunStatus.PENDING, # Initial status
            created_at=datetime.utcnow()
        )
        db.add(db_run)
        await db.commit()
        await db.refresh(db_run)
        return db_run

    async def get_run(self, db: AsyncSession, run_id: uuid.UUID) -> Optional[RunORM]:
        """
        Fetches a specific run by its ID.

        Args:
            db: The AsyncSession for database interaction.
            run_id: The UUID of the run to fetch.

        Returns:
            The RunORM object if found, otherwise None.
        """
        result = await db.execute(select(RunORM).where(RunORM.id == run_id))
        return result.scalars().first()

    async def update_run_status(
        self,
        db: AsyncSession,
        run_id: uuid.UUID,
        status: RunStatus,
        output_data: Optional[Dict[str, Any]] = None
    ) -> Optional[RunORM]:
        """
        Updates the status and potentially the output of a run.
        Sets started_at or ended_at timestamps based on the status.

        Args:
            db: The AsyncSession for database interaction.
            run_id: The UUID of the run to update.
            status: The new RunStatus.
            output_data: The output data (for COMPLETED) or error info (for FAILED).

        Returns:
            The updated RunORM object if found and updated, otherwise None.
        """
        run = await self.get_run(db, run_id)
        if not run:
            return None # Or raise an exception

        run.status = status
        now = datetime.utcnow()

        if status == RunStatus.RUNNING and not run.started_at:
            run.started_at = now
        elif status in [RunStatus.COMPLETED, RunStatus.FAILED, RunStatus.CANCELLED]:
            if not run.started_at: # Should have been set when RUNNING, but set JIC
                 run.started_at = now
            run.ended_at = now
            run.output_data = output_data # Store output/error

        db.add(run) # Add the modified object to the session
        await db.commit()
        await db.refresh(run)
        return run

# You might want a singleton instance or use dependency injection
run_service = RunService()

File: frontend/src/api/runs.ts
import apiClient from './client';

// Define the RunStatus enum matching the backend
export enum RunStatus {
  PENDING = "pending",
  RUNNING = "running",
  COMPLETED = "completed",
  FAILED = "failed",
  CANCELLED = "cancelled",
}

// Define the structure of a Run based on the backend model
export interface Run {
  id: string; // UUID as string
  status: RunStatus;
  created_at: string; // Datetime as ISO string
  started_at: string | null; // Optional datetime as ISO string or null
  ended_at: string | null; // Optional datetime as ISO string or null
  runnable_id: string; // UUID as string
  runnable_type: 'agent' | 'team'; // Constrained string type
  input_variables: Record<string, any> | null; // Dict as Record or null
  output_data: Record<string, any> | null; // Dict as Record or null
}

// Interface for creating a run (matches RunCreate -> RunBase)
export interface RunCreateData {
  runnable_id: string;
  runnable_type: 'agent' | 'team';
  input_variables?: Record<string, any> | null;
}

// Function to fetch all runs (with optional filters)
export const fetchRuns = async (params?: { runnable_id?: string; status?: RunStatus; limit?: number; skip?: number }): Promise<Run[]> => {
  try {
    // Pass filters as query parameters
    const response = await apiClient.get<Run[]>('/runs', { params });
    return response.data;
  } catch (error) {
    console.error("Error fetching runs:", error);
    throw error;
  }
};

// Function to fetch a single run by ID
export const fetchRunById = async (id: string): Promise<Run> => {
  try {
    const response = await apiClient.get<Run>(`/runs/${id}`);
    return response.data;
  } catch (error) {
    console.error(`Error fetching run with ID ${id}:`, error);
    throw error;
  }
};

// Function to create/start a new run
export const createRun = async (runData: RunCreateData): Promise<Run> => {
  try {
    const response = await apiClient.post<Run>('/runs', runData);
    return response.data;
  } catch (error) {
    console.error("Error creating run:", error);
    throw error;
  }
};

// No update/delete endpoints currently defined in the backend for runs
